demos/common/monitors/memory_monitor.cpp

```diff
@@ -0,0 +1,258 @@
+// Copyright (C) 2019 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "memory_monitor.h"
+
+#ifdef WIN32
+#include <algorithm>
+
+#define PSAPI_VERSION 1 // for psapi
+#include <windows.h>
+#include <psapi.h>
+
+namespace {
+double getMemTotal() {
+    PERFORMANCE_INFORMATION performanceInformation;
+    if (!GetPerformanceInfo(&performanceInformation, sizeof(performanceInformation))) {
+        throw std::runtime_error("GetPerformanceInfo() failed");
+    }
+    return static_cast<double>(performanceInformation.PhysicalTotal * performanceInformation.PageSize)
+        / (1024 * 1024 * 1024);
+}
+double getSwapTotal() {
+    PERFORMANCE_INFORMATION performanceInformation;
+    if (!GetPerformanceInfo(&performanceInformation, sizeof(performanceInformation))) {
+        throw std::runtime_error("GetPerformanceInfo() failed");
+    }
+    return static_cast<double>(performanceInformation.CommitLimit * performanceInformation.PageSize)
+        / (1024 * 1024 * 1024);
+}
+}
+
+MemoryMonitor::MemoryMonitor() :
+    memTotal{getMemTotal()},
+    swapTotal{getSwapTotal()},
+    enabled{false},
+    samplesNumber{0},
+    meanMem{0},
+    meanSwap{0},
+    maxMem{0},
+    maxSwap{0} {}
+
+bool MemoryMonitor::isEnabled() const {
+    return enabled;
+}
+
+void MemoryMonitor::enable(std::size_t historySize) {
+    this->historySize = historySize;
+    enabled = true;
+}
+
+void MemoryMonitor::disable() {
+    enabled = false;
+}
+
+void MemoryMonitor::collectData() {
+    PERFORMANCE_INFORMATION performanceInformation;
+    if (!GetPerformanceInfo(&performanceInformation, sizeof(performanceInformation))) {
+        throw std::runtime_error("GetPerformanceInfo() failed");
+    }
+    double usedMem = static_cast<double>(
+        (performanceInformation.PhysicalTotal - performanceInformation.PhysicalAvailable)
+        * performanceInformation.PageSize) / (1024 * 1024 * 1024);
+    double usedSwap = static_cast<double>(performanceInformation.CommitTotal * performanceInformation.PageSize)
+        / (1024 * 1024 * 1024);
+
+    if (0 == samplesNumber) {
+        maxMem = meanMem = usedMem;
+        maxSwap = meanSwap = usedSwap;
+        samplesNumber = 1;
+    } else {
+        meanMem = (meanMem * samplesNumber + usedMem) / (samplesNumber + 1);
+        meanSwap = (meanSwap * samplesNumber + usedSwap) / (samplesNumber + 1);
```

Comment:
It seems a bit pointless (and potentially precision-destroying) to keep multiplying and dividing by the number of samples. Why not keep the sum and only divide it by the number of samples when you need the mean?

Reply:
updated